docs/source/transactions/life_of_a_transaction.rst

```diff
@@ -0,0 +1,179 @@
+The transaction task must precede committing, as that opens with constraint checking which needs to know what writes
+were performed.
+
+Commit
+======
+
+Committing a write transaction is a multi-stage process:
+
+1. Check constraints if applicable.
+2. Lock rows or cells (depending on the table conflict handler) for a given table.
+3. Check for write/write conflicts.
+4. Persist information about writes to the targeted sweep queue.
+5. Persist key-value pairs that were written to the database.
+6. Get the commit timestamp.
+7. For serializable transactions, check that the state of the world at commit time is same as that at start time.
+   This involves re-executing the relevant read queries we made and checking that we read the same *values*
+   (note that ABA situations where the state of the world changed and changed back are permitted).
+8. Verify that locks are still valid.
+9. Verify user-specified pre-commit conditions (if applicable).
+10. Atomically putUnlessExists into the transactions table.
+
+After we have successfully written to the transactions table, the transaction is considered committed.
+
+The ordering of these steps is important:
+
+1. Checking constraints must be done before committing. These constraints do not depend on the rows being locked,
+   so we should do this before acquiring locks (generally we want to minimise critical sections).
+2. If we check for conflicts before acquiring locks, it is possible that another transaction may write to a value
+   we've written to underneath us. This should cause a write-write conflict, but we will miss it. Taking locks first
+   means that this is only possible if we've lost our locks before the check - but if that is the case then we will
+   not commit as we will fail when we check our locks (step 8).
+3. Checking for write/write conflicts must be done before checking locks are still valid (step 8), as we could
+   otherwise lose the lock and have a thorough sweep clear out all evidence of spanning/dominating writes.
+   It may be possible to postpone this, though we would be doing unnecessary work.
+4. If we write to the database before writing to the targeted sweep queue, we may write values to the database and
+   then crash. We then have no knowledge of these uncommitted values, meaning that we'll permanently have cruft
+   lying around. It's possible that Background Sweep can be used to clear any such values, but that may take a very
+   long time.
+5. If we get the commit timestamp before we write key-value pairs to the database, another transaction could start
+   after we commit, and may read cells that we have written to. Our own writes may not have been made, but the
+   other transaction must see them.
+6. The serializable commit check requires us to know the commit timestamp.
+7. For conservatively swept tables, the ordering of this step and checking locks is not critical (consider that our
```

Comment:
Hard to know what 'this step' refers to, since #6 above was already about the serializable check (or at least it's worded such that it seems more like a constraint on the serializable check rather than the commit timestamp). It still maps to #7, but I scrolled up when reviewing to check what it referred to.

Reply:
Yep. I should reference all steps here very explicitly by number. My intention was that each of the things in this list explains why step `N` must happen before `N+1` (or it doesn't), but this can indeed be confusing.